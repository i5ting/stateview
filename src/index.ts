
// 

declare global {
    interface Window {
        stateview: any
    }
}

export * from './type'
export * from './stateview'
export * from './layer'
export * from './setViewState'
export * from './state'
export * from './debug' 