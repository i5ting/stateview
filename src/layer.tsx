import React, { Fragment } from 'react';

interface LayerProps {
    state: string;
    component?: any;
    children?: any;
    data?: any;
}

export function Layer(props: LayerProps) {
    return <Fragment>{props.component}</Fragment>;
}
