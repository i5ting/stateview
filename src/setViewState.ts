/**
 * 封装stateview.show和stateview.datashow
 */

function _show(instance: any, state: string, data?: any) {
    if (!data) {
        instance.show(state)
    } else {
        instance.datashow(state, data)
    }
}

export const setViewState = (state: string, data?: any) => {
    if (!data) {
        _show(window.stateview, state)
    } else {
        _show(window.stateview, state, data)
    }
};

export const getStateview = (group: string) => {
    if (group === undefined)
        throw new Error("there is no group arg in getStateview method")

    let _stateview = window.stateview[group]

    return Object.assign(_stateview, {
        setViewState: (state: string, data?: any) => {
            if (!data) {
                _show(_stateview, state)
            } else {
                _show(_stateview, state, data)
            }
        }
    })
};
