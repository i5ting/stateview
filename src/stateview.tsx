import React, { Fragment, useLayoutEffect, useState } from 'react';
import { Debug } from './debug';
import { getQueryStringByName } from './utils';
import { SState } from './state';

const debug = Debug("stateview.jsx")

var count: number = 0;
var GlobalStateMapping: any = {}
var groups: any = {}

export const Stateview = (props: any) => {
    const [visibaleComponent, setVisibaleComponent] = useState(0);

    let states: any = []
    let group: string = props.group
    let hasGroupName: Boolean = (group !== undefined)
    let instance: any = hasGroupName ? GlobalStateMapping[group] : GlobalStateMapping

    if (!instance) instance = {}
    if (hasGroupName) {
        debug('group name = ' + group)
        groups[group] = true
    }
    // states.for()
    React.Children.forEach(props.children, child => {
        // const childType = { ...child.type }
        debug(child)
        // console.log(child.name)
        debug(child.props.state)
        states.push(child.props.state)

        let component = child.props.component

        if (!child.props.component) {
            component = child.props.children
        }

        instance[child.props.state] = {
            show: setVisibaleComponent,
            child: child,
            component: component
        }
    })

    const sState = SState(GlobalStateMapping)

    if (count === 0) {
        GlobalStateMapping.currentState = props.default
        window.stateview = sState
    }

    if (hasGroupName) {
        window.stateview[group] = SState(instance)
    }

    debug(count++)
    debug(sState)

    useLayoutEffect(() => {
        // show default
        let d = instance[props.default]
        debug("useLayoutEffect d")
        debug(d)
        let _component = d.component
        if (props.data) {
            _component = React.cloneElement(
                d.component,
                {
                    data: props.data
                }
            )
        }
        d.show(_component)
    }, [])

    let i = getQueryStringByName('s')
    debug(props.children[i])

    return (
        <Fragment >
            {visibaleComponent}
        </Fragment>
    );
}
