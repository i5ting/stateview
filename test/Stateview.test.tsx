import { shallow, mount } from 'enzyme';
import React from 'react';
// import { act } from 'react-dom/test-utils';
import { Stateview, Layer, setViewState } from '../src';
// import type { ExtendablePromise } from '../src/types';

const Example1 = () => {

    function unlogin() {
        setViewState('unlogin')
    }

    function logined() {
        setViewState('logined')
    }

    return (
        <Stateview default='unlogin'>
            <Layer state='logined'>
                <h1>Logined, <button id="go_to_unlogin_button" onClick={unlogin}>go to UnLogin</button></h1>
            </Layer>
            <Layer state='unlogin'>
                <h1 >UnLogin, <button id="go_to_logined_button" onClick={logined}>go to Logined</button></h1>;
            </Layer>
        </Stateview>
    );
}

describe('<Stateview /> component', () => {
    describe('Truthy cases', () => {
        describe('Stateview', () => {
            test('GIVEN <Stateview /> render default Layer', () => {
                const wrapped = shallow(
                    <Example1 />
                );

                expect(wrapped).toMatchSnapshot();
                expect(wrapped.containsMatchingElement(<h1>UnLogin, <button>go to Logined</button></h1>)).toBe(true);
            });

            test('GIVEN <Stateview /> click button to switch Layer', () => {
                const wrapped = mount(
                    <Example1 />
                );

                wrapped.find('button').simulate('click');
                expect(wrapped.containsMatchingElement(<h1>Logined, <button>go to UnLogin</button></h1>)).toBe(true);
                wrapped.find('button').simulate('click');
                expect(wrapped.containsMatchingElement(<h1>UnLogin, <button>go to Logined</button></h1>)).toBe(true);
            });
        });

    });
});
