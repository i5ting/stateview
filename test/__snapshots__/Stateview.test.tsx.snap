// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stateview /> component Truthy cases Stateview GIVEN <Stateview /> render default Layer 1`] = `
<Stateview
  default="unlogin"
>
  <Layer
    state="logined"
  >
    <h1>
      Logined, 
      <button
        id="go_to_unlogin_button"
        onClick={[Function]}
      >
        go to UnLogin
      </button>
    </h1>
  </Layer>
  <Layer
    state="unlogin"
  >
    <h1>
      UnLogin, 
      <button
        id="go_to_logined_button"
        onClick={[Function]}
      >
        go to Logined
      </button>
    </h1>
    ;
  </Layer>
</Stateview>
`;
